### DISCUSSION TITLE
Blazor app - lock on AssignIdentity when inserting

### DISCUSSION THREAD

**Discussion by tompru** [JasperFx/marten#2161] [General] [answered]
Hi!

Currently I'm developing Blazor Server Side app using Marten as document database and event store (awesome tool!).
I'am facing problem with lock when I'm trying to insert document with Id of type 'int'. If I am correct - this problem occures only on Blazor apps. 

Way to reproduce:
1: Create new Blazor Server Side app
2. Add Marten
3. Execute some read query using Marten
4. Try to execute insert

Result should be lock on invoking session.Insert(object)

I have debugged Marten internals and I found out that lock happens in MartenDatabase.cs on method resetSequences (generateOrUpdateFeature).
It looks like Lazy initialization of _sequences on Blazor apps couses problems. When I have been debugging I tried to remove lazy initialization and it solved the problem for me, but I don't know if that is acceptable solution since (if I am correct) _sequences are needed only for HiLoIdGeneration and IdentityKeyGeneration.

Also I found workaround for this problem. If first operation you invoke using Marten is session.Insert(object) - then _sequences are initialized and lock doesn't occur anymore in blazor apps.

It looks like generally on Blazor apps, method generateOrUpdateFeature couses some problems with locks - similar issue:
https://github.com/JasperFx/marten/issues/2030

**Comment by jeremydmiller** [comment, marked as answer]
You can somewhat easily work around this w/ the new 5.0 release using the ApplyAllChangesOnStartup(). See maybe this blog post?

https://jeremydmiller.com/2022/03/30/improving-the-development-and-production-time-experience-with-marten-v5/

**Comment by tompru** [comment]
Thanks! :) This work around with ApplyAllChangesOnStartup() solved my problem with no need to do some dirty hacks :D
